mdm/mw/udidauth.go

```diff
@@ -0,0 +1,86 @@
+package mw
+
+import (
+	"bytes"
+	"context"
+	"crypto/x509"
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	"github.com/micromdm/micromdm/mdm"
+	devicebuiltin "github.com/micromdm/micromdm/platform/device/builtin"
+)
+
+func UDIDAuthMiddleware(store *devicebuiltin.DB) mdm.Middleware {
+	return func(next mdm.Service) mdm.Service {
+		return &udidAuthMiddleware{
+			store: store,
+			next:  next,
+		}
+	}
+}
+
+type udidAuthMiddleware struct {
+	store *devicebuiltin.DB
+	next  mdm.Service
+}
+
+func (mw *udidAuthMiddleware) checkCertUDIDMatch(cert *x509.Certificate, udid string) (bool, error) {
+	devByUUID, err := mw.store.DeviceByUDID(udid)
+	if err != nil {
+		if devicebuiltin.IsNotFound(err) {
+			return false, errors.New("device not found; no certificate UDID match")
+		}
+		return false, err
+	}
+
+	if devByUUID.DeviceCert == nil {
+		// nil certificate. this could be a potentially dangerous thing
+		// to allow but if, truely, there is no certificate then the
+		// most likely thing is that this is a "legacy" device that
+		// needs its certificate updated in the database.
+		//
+		// but for testing simply fail out.
+
+		// upgrade any devices that don't have a certificate to have one
+		devByUUID.DeviceCert = cert
+		err = mw.store.Save(devByUUID)
+		if err != nil {
+			return false, err
+		}
+		fmt.Println("SAVED NEW!")
+		return true, nil
+	}
+	if bytes.Compare(cert.Raw, devByUUID.DeviceCert.Raw) == 0 {
```

Comment:
`bytes.Compare` is not suitable for security checks because it's sensitive to timing attacks. https://golang.org/pkg/crypto/subtle/#ConstantTimeCompare is a good substitute.